### PR TITLE
Custom Directive Implementations

### PR DESCRIPTION
**Pull request by mathewbyrne**
This adds a new configuration option for a user to bypass codegen for a directive and provide their own implementation like so:

```yml
directives:
  exampleDirective:
    implementation: "github.com/99designs/gqlgen.ExampleDirective"
```

This feature is primarily intended to support plugins (#228) where a plugin could provide an implementation during codegen and avoid it being added to the `DirectiveRoot`.

This is currently a pretty bare bones implementation, so I'd welcome any feedback.  One major question is how do we provide run-time dependencies to custom implementations?  Also is there a way we can provide a no-op for plugins that don't need a runtime implementation for a directive?

**TODO**
- [x] Validate implementation signature
- [x] Better test coverage

### CODE REVIEW

**Comment by JulienBreux** [comment]
Nice work. Maybe need test?

**Comment by mathewbyrne** [comment]
Yep, there's a couple of todos left on this at the moment; I thought I'd get it up for any feedback first.

File: codegen/templates/generated.gotpl [hunk 1]
```diff
@@ -226,6 +226,20 @@ func (ec *executionContext) FieldMiddleware(ctx context.Context, obj interface{}
 		switch d.Name {
 		{{- range $directive := .Directives }}
 		case "{{$directive.Name}}":
+			{{- if $directive.Implementation }}
```
Comment (vektah) [inline codegen/templates/generated.gotpl]:
```
{{- with $directive.Implementation}}
    // args
    {{ . }}
{{-end}}
```
?

also the indenting went a bit funky here, aim for readable templates, gofmt will clean the code up.

**Comment by mathieupost** [comment]
Is this still possible? Cant seem to find documentation about it and if I try to use it like this 
```
directives:
  exampleDirective:
    implementation: "github.com/99designs/gqlgen.ExampleDirective"
```
I'm getting 
```
unable to parse config: yaml: unmarshal errors:
  line 41: field implementation not found in type config.DirectiveConfig
```

**Pull request merged**